### ISSUE TITLE
Complex Numbers Support

### ISSUE THREAD

**Issue by anjali411** [pytorch/pytorch#33152]
Previous discussion: https://github.com/pytorch/pytorch/issues/755

High Priority
----------------------------------------------------------------------------------------------------------- 

**Operator support**
- [x] [New operator] [`torch.linalg.adjoint()`](https://github.com/pytorch/pytorch/issues/45063)
- [x] [`torch.linalg.eig` (both CPU and CUDA, waiting on TH to ATen port)](https://github.com/pytorch/pytorch/issues/43081)
- [ ] Add kernel for vectorization on CPU for `torch.complex` and `torch.polar`
- [ ] `torch.arange`
- [x] [Deprecate the current linspace, logspace automatic dtype inference and get rid of the unnecessary warning ](https://github.com/pytorch/pytorch/issues/53171)
- [x] Conjugate transpose variants of linear algebra operators for e.g., see here an issue for [`torch.triangular_solve`](https://github.com/pytorch/pytorch/issues/45734)
- [ ] [`torch.special`](https://github.com/pytorch/pytorch/issues/53440)

`TH to ATen ports that would greatly help unblock some of the work listed here`:

- [x] `mode` #24731, #24526
- [x] `renorm` #24616
- [x] `put_` #24751, #24614
- [x] `take` #24640, #24772
- [x] `nonzero (CPU)` #24745
- [x] `index_fill_` #24577, #24714
- [x] `index_copy` #24523, #24670
- [x] `masked_scatter`
- [x] `masked_fill`  #49543

**Complex support for real valued `torch.nn` loss functions**
Tasks listed on https://github.com/pytorch/pytorch/issues/46642 along with a sample PR for `torch.nn.L1Loss`.

**Complex Autograd supported but untested:**
- [x] [`torch.dist`](https://github.com/pytorch/pytorch/issues/53516)
- [x] [`torch.masked_fill`](https://github.com/pytorch/pytorch/issues/53523)
- [x] [`torch.Tensor.masked_scatter_`](https://github.com/pytorch/pytorch/issues/53608)
- [ ] `torch.Tensor.scatter_`

**Enhance testing**

- [x] Migrate the op tests (shown here: https://github.com/pytorch/pytorch/blob/a9e4bb56e54e83a69fbe7fad6d637817db1dd6a4/test/test_autograd.py#L5099-L5122) from method_tests to `OpInfo` based tests. After this migration is complete, remove the method_test generation logic for complex types: https://github.com/pytorch/pytorch/blob/a9e4bb56e54e83a69fbe7fad6d637817db1dd6a4/test/test_autograd.py#L5131-L5313
- [ ] Add `R->C` tests for complex functions. (contact @anjali411 if you'd like to work on this task)

**Complex Autograd support for the following ops supported for complex:**

- [x] [`torch.sigmoid`](https://github.com/pytorch/pytorch/issues/48552)
- [x] [`torch.eig`](https://github.com/pytorch/pytorch/issues/51621#issuecomment-777115794)
- [x] [`torch.lu`](https://github.com/pytorch/pytorch/issues/52891)
- [x] [`torch.cross`](https://github.com/pytorch/pytorch/issues/53512) (Also add tests for function correctness)
- [x] [`torch.cumsum`](https://github.com/pytorch/pytorch/issues/53182)
- [x] [`torch.cumprod`](https://github.com/pytorch/pytorch/issues/53183)
- [x] [`torch.linalg.det`](https://github.com/pytorch/pytorch/issues/53513)
- [x] [`torch.Tensor.index`](https://github.com/pytorch/pytorch/issues/53605)
- [x] [`torch.lerp`](https://github.com/pytorch/pytorch/issues/53606)
- [x] [`torch.prod`](https://github.com/pytorch/pytorch/issues/53511)
- [x] (waiting on TH to ATen port)`torch.Tensor.put_`
- [x] [`rsub`](https://github.com/pytorch/pytorch/issues/53643)
- [x] [`torch.symeig`](https://github.com/pytorch/pytorch/issues/53651)
- [x] [`index_put`](https://github.com/pytorch/pytorch/issues/53645)
- [x] [`torch.unfold`](https://github.com/pytorch/pytorch/issues/51875)
- [x] `torch.diag`
- [x] [`torch.Tensor.masked_fill_`](https://github.com/pytorch/pytorch/pull/51281)
- [x] [`torch.trace`](https://github.com/pytorch/pytorch/pull/51537)
- [x] `torch.polar`

-------------------------------------------------------------------------------------
**Linear Algebra Ops**

- [x] torch.logdet
- [ ] torch.trapz
- [ ] torch.pca_lowrank
- [ ] torch.svd_lowrank
- [ ] torch.ormqr
- [x] [torch.orgqr](https://github.com/pytorch/pytorch/pull/50502)
- [x] torch.matrix_power
- [x] [`torch.cholesky_inverse`](https://github.com/pytorch/pytorch/pull/50269)
- [x] `torch.chain_matmul`
- [x] `torch.ger / torch.outer`
- [x] [torch.solve](https://github.com/pytorch/pytorch/issues/41084) [CUDA]
- [x] [torch.bmm [CUDA, waiting on TH to ATen port]
](https://github.com/pytorch/pytorch/pull/42553)
- [x] [torch.lu_solve](https://github.com/pytorch/pytorch/pull/46862)
- [x] [torch.cholesky_solve [CUDA]](https://github.com/pytorch/pytorch/pull/47047)
- [x] torch.baddbmm

Other ops:

- [x] `torch.isnan` and `torch.isfinite`
- [x] [`torch.linspace`](https://github.com/pytorch/pytorch/issues/38775) 
- [x] [`torch.logspace`](https://github.com/pytorch/pytorch/issues/38779)
- [x] [torch.trace](https://github.com/pytorch/pytorch/issues/33152#issuecomment-714611987)
- [x] [torch.mean](https://github.com/pytorch/pytorch/pull/47048)
- [x] [ torch.dot [CPU and CUDA]](https://github.com/pytorch/pytorch/pull/42745)
- [x] [torch.vdot](https://github.com/pytorch/pytorch/issues/42747)
- [x] [torch.bmm](https://github.com/pytorch/pytorch/pull/42383) [CPU]
- [x] [torch.symeig [CPU and CUDA]](https://github.com/pytorch/pytorch/issues/45061)
- [x] torch.addmv (added [here](https://github.com/pytorch/pytorch/pull/37924) and [here](https://github.com/pytorch/pytorch/pull/40238))
- [x] [torch.cholesky [CUDA]](https://github.com/pytorch/pytorch/issues/44637)
- [x] [torch.svd [CUDA]](https://github.com/pytorch/pytorch/pull/45795)
- [x] [torch.qr [CUDA]](https://github.com/pytorch/pytorch/pull/45032)
- [x] [torch.lu, torch.lu_unpack [CUDA]](https://github.com/pytorch/pytorch/pull/45898)
- [x] [torch.det](https://github.com/pytorch/pytorch/pull/45980)
- [x] [torch.inverse [CUDA]](https://github.com/pytorch/pytorch/pull/45034)
- [x] [torch.triangular_solve [CUDA]](https://github.com/pytorch/pytorch/pull/46916)

**Spectral Op Migration:** https://github.com/pytorch/pytorch/issues/42175

**complex views**

- [x] [tensor.real](https://github.com/pytorch/pytorch/issues/35494)

- [x] [tensor.imag](https://github.com/pytorch/pytorch/issues/35494) 
- [x] [torch.view_as_real](https://github.com/pytorch/pytorch/pull/39099)

- [x] [torch.view_as_complex](https://github.com/pytorch/pytorch/pull/39099) 

**[Factory Functions]**

- [x] [torch.complex](https://github.com/pytorch/pytorch/issues/38458#issuecomment-636066256) , torch.polar [35312](https://github.com/pytorch/pytorch/issues/35312)
            
- [x] torch.rand

- [x] torch.randn

- [x] torch.from_numpy(complex_array) ~~blocked on: https://github.com/pytorch/pytorch/issues/24759 https://github.com/pytorch/pytorch/issues/24623~~

Other functions:

- [x] [torch.sgn](https://github.com/pytorch/pytorch/issues/36655) 
- [x] [torch.index_select [CUDA]](https://github.com/pytorch/pytorch/issues/24578#issuecomment-630521691)
- [x] [Disabled] [torch.remainder](https://github.com/pytorch/pytorch/issues/34266)
- [x] [Disabled] [torch.clamp](https://github.com/pytorch/pytorch/issues/33568) 

**Complex Number Support for `torch.nn.distributed`**: https://github.com/pytorch/pytorch/issues/45760
 
**Complex Autograd Guide:**

To understand and obtain the formula for complex derivatives, check out: https://pytorch.org/docs/master/notes/autograd.html#what-are-complex-derivatives 

The list of operators for which complex autograd is supported and tested can be found here: https://github.com/pytorch/pytorch/blob/72bc3d9de4f8baa28ebc8f9399ad86cefe89d9f5/tools/autograd/gen_variable_type.py#L154. If an operator has an entry in the file `derivatives.yaml`, then to enable complex backward for that operator, it would have to be added to [`GRADIENT_IMPLEMENTED_FOR_COMPLEX`](https://github.com/pytorch/pytorch/blob/b61671ccd2f8664c306ccd661e030d8762cca693/tools/autograd/gen_variable_type.py#L158) in `gen_variable_type.py`.

To run a common_methods test for complex, add an entry here: https://github.com/pytorch/pytorch/blob/b61671ccd2f8664c306ccd661e030d8762cca693/test/test_autograd.py#L4927
To run a common_methods test only for complex, add an entry here: https://github.com/pytorch/pytorch/blob/b61671ccd2f8664c306ccd661e030d8762cca693/test/test_autograd.py#L4920

**Autograd tasks:**

- [x] Gradcheck logic for `C -> C, C -> R, R -> C` functions. (#43208 )

- [x] Disable complex autograd for operators not tested for complex. (#45461 )

- [x] [scalar mul backward](https://github.com/pytorch/pytorch/issues/43349)

- [ ] [CUDA complex acos fails gradcheck on Windows](https://github.com/pytorch/pytorch/issues/43852)

- [x] [Update pow_backward](https://github.com/pytorch/pytorch/issues/41510)

Other tasks:

- [x] type promotion (eg. Float/Long/... tensor + a complex number = complex tensor) 33780 [https://github.com/pytorch/pytorch/issues/33780]
- [x] Python API for ComplexFloat and ComplexDouble Storage 
- [x] Disable complex sign (see #36323)
- [x] Disable comparison ops for complex(clamp, sort, min, max...) (see #36374, #33568, and #36444)
- [x] Implement isclose for complex (see #36462, #36029, and #35471) [36456](https://github.com/pytorch/pytorch/pull/36456)
- [x] Complex Printing for scientific notation [38285](https://github.com/pytorch/pytorch/issues/38285) [33494](https://github.com/pytorch/pytorch/issues/33494)
- [x] testing in test_torch.py: https://github.com/pytorch/pytorch/issues/36029
- [x] conj() is not supported on real cuda tensors

**Discussions:**
1. [Comparison ops for complex numbers](https://github.com/pytorch/pytorch/issues/36444)

**Numpy parity:**

- [x] Inconsistency between torch.abs and np.abs [33567](https://github.com/pytorch/pytorch/issues/33567)

- [x] [torch.angle is divergent from numpy.angle](https://github.com/pytorch/pytorch/issues/41240)

- [ ] [torch.abs(complex) is divergent from NumPy on vectorized NaN values](https://github.com/pytorch/pytorch/issues/41246)

c10::complex tracker: https://github.com/pytorch/pytorch/issues/35284#issuecomment-605350370

cc @ezyang @anjali411 @dylanbespalko @mruberry

**Comment by vadimkantorov** [comment]
conj() is not currently supported on real cuda tensors ( https://github.com/pytorch/pytorch/pull/29488/files#diff-e43d1ac289e6d25c6db003f0b2ebf488R67):

```
>>> a = torch.zeros(3, device = 'cuda')
>>> a.conj()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Could not run 'aten::conj.out' with arguments from the 'CUDATensorId' backend. 'aten::conj.out' is only available for these backends: [CPUTensorId, VariableTensorId].
>
```

On real CPU tensors it works

**Comment by vadimkantorov** [comment]
Also in 1.4.0:
```
>>> a = torch.zeros(3, dtype = torch.complex128, device = 'cuda')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Could not run 'aten::empty.memory_format' with arguments from the 'ComplexCUDATensorId' backend. 'aten::empty.memory_format' is only available for these backends: [CUDATensorId, SparseCPUTensorId, VariableTensorId, CPUTensorId, MkldnnCPUTensorId, SparseCUDATensorId].
>
```

**Comment by anjali411** [comment]
> a = torch.zeros(3, dtype = torch.complex128, device = 'cuda')

I fixed this last week. it was a printing issue. should work now :D

**Comment by anjali411** [comment]
> on 1.4.0 conj() is not supported on real cuda tensors ( https://github.com/pytorch/pytorch/pull/29488/files#diff-e43d1ac289e6d25c6db003f0b2ebf488R67):
> 
> ```
> >>> a = torch.zeros(3, device = 'cuda')
> >>> a.conj()
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> RuntimeError: Could not run 'aten::conj.out' with arguments from the 'CUDATensorId' backend. 'aten::conj.out' is only available for these backends: [CPUTensorId, VariableTensorId].
> >
> ```
> 
> On real CPU tensors it works

added to the list

**Comment by vadimkantorov** [comment]
@pearu A few complex numbers quirks from your SVD PR hopefully will be fixed soon :)

**Comment by dylanbespalko** [comment]
As you may know, the CPU and GPU is not kind towards natively supporting hardware acceleration for complex numbers. On the CPU and GPU you need to call predefined kernels that implemented by the processor manufacturer.

Alternatively, the FPGA using Xilinx Vitis allows ML engineers to define they're own kernels. If you are concerned by the performance degradation between complex and real numbers on the CPU, I encourage you to read my latest blogs:
[pytorch-for-fpga-part-1-heterogeneous-processing](https://www.dylan-bespalko.scikit-nonlinear.org/post/pytorch-for-fpga-part-1-heterogeneous-processing)
[pytorch-for-fpga-part-2-basic-fpga-optimizations](https://www.dylan-bespalko.scikit-nonlinear.org/post/pytorch-for-fpga-part-2-basic-fpga-optimizations)
[pytorch-for-fpga-part-3-advanced-fpga-optimizations](https://www.dylan-bespalko.scikit-nonlinear.org/post/pytorch-for-fpga-part-3-advanced-fpga-optimizations)
[pytorch-for-fpga-part-4-deploying-pytorch-kernels](https://www.dylan-bespalko.scikit-nonlinear.org/post/pytorch-for-fpga-part-4-deploying-pytorch-kernels)
pytorch-for-fpga-part-5-complex-and-quaternion-numbers (Coming soon).

I am posting this early because I really need your support with an issue on the Xilinx FPGA. Please vote up this issue on the [Xilinx Forum](https://forums.xilinx.com/t5/AI-and-Vitis-AI/Using-WideType-Data-Container-to-Increase-Memory-Bit-Width/td-p/1078666) and [Github](https://github.com/Xilinx/Vitis-AI/issues/46).

**Comment by vadimkantorov** [comment]
It seems that pointwise complex multiplication is implemented, right? Then I can close https://github.com/pytorch/pytorch/issues/6514

Also a question: will torch.rfft / irfft start returning complex tensors by default (a first step may be a dtype argument). My usecase: https://gist.github.com/vadimkantorov/d9b56f9b85f1f4ce59ffecf893a1581a

**Comment by ezyang** [comment]
Yeah, pointwise complex multiply is all done. But if you have an old fashioned float tensor it's not possible to conveniently call the kernel due to #33950

Eventually we want rfft/irfft to return complex by default but this has BC implications which we haven't studied yet. An argument of some sort is definitely a good first step.

**Comment by vadimkantorov** [comment]
@ezyang Does real() and imag() always copy? In many circumstances for advanced complex formulas it may be very useful to "unbind" complex tensor into real/imag without copying. I proposed `copy=False` argument for conj() in https://github.com/pytorch/pytorch/issues/30237 but is rejected for now

**Comment by ezyang** [comment]
Yes right now it always copies. It would be good to avoid the copy but #33950 needs to happen first.

**Comment by VIGNESHinZONE** [comment]
Could i also help in contributing and can someone pls guide me

**Comment by krshrimali** [comment]
> Also in 1.4.0 (maybe alread fixed in master):
> 
> ```
> >>> a = torch.zeros(3, dtype = torch.complex128, device = 'cuda')
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> RuntimeError: Could not run 'aten::empty.memory_format' with arguments from the 'ComplexCUDATensorId' backend. 'aten::empty.memory_format' is only available for these backends: [CUDATensorId, SparseCPUTensorId, VariableTensorId, CPUTensorId, MkldnnCPUTensorId, SparseCUDATensorId].
> 
> >>> a = torch.zeros(3, dtype = torch.complex128, device = 'cpu')
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> RuntimeError: Could not run 'aten::empty.memory_format' with arguments from the 'ComplexCPUTensorId' backend. 'aten::empty.memory_format' is only available for these backends: [CUDATensorId, SparseCPUTensorId, VariableTensorId, CPUTensorId, MkldnnCPUTensorId, SparseCUDATensorId].
> ```

Was this fixed for CPU devices as well (or do we not support complex tensors on CPU)? I get this error:

```bash
>>> a = torch.zeros(3, device='cpu', dtype=torch.complex64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Complex dtype not supported.
```

Posting this here, as this discussion seemed relevant.
cc: @anjali411

**Comment by anjali411** [comment]
@krshrimali Hey Kushashwa! Yeah that's expected. Complex tensor construction was turned off in the past release(s) and will be off until we plan to release it.

**Comment by krshrimali** [comment]
> @krshrimali Hey Kushashwa! Yeah that's expected. Complex tensor construction was turned off in the past release(s) and will be off until we plan to release it.

Sorry, I took a better look at it and rebuilt using master branch - and it worked (though, as you say it is not in the stable release), so I deleted the message. For those who land here looking for the message, it was about whether complex types are supported for CPU tensors (see below):

```bash
>>> x = torch.tensor(1, dtype=torch.complex64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Complex dtype not supported
```

Thanks, @anjali411 for your response!

**Comment by vadimkantorov** [comment]
One important design question: are complex-specific attributes like `.real`, `.imag` (with real and imag supporting copy-based assignment too), `.conj()`, `.phase()` are going to be supported only on complex types? What is going to be the real/complex abstraction/polymorphism theory?

Many of optimization solvers may have same path if complex-specific attributes are efficiently supported, otherwise they must have some convenient concise ways of checking the type. Then a question is how it is going to be traced.

**Comment by ezyang** [comment]
Right now, we're generally following what would be needed for numpy compatibility. I don't know off the top of my head what happens, e.g., with assignment, but that will give a basic answer to this question, although it may result in less than satisfactory results with optimizations.

**Comment by Wenhao-Jin** [comment]
> > a = torch.zeros(3, dtype = torch.complex128, device = 'cuda')
> 
> I fixed this last week. it was a printing issue. should work now :D

Hi, I also got the error: 
RuntimeError: Could not run 'aten::conj.out' with arguments from the 'CUDATensorId' backend. 'aten::conj.out' is only available for these backends: [CPUTensorId, VariableTensorId].

I'm using version  pytorch-geometric 1.4.0. Would you mind also fixing this in 1.4.0? It's because I only have cuda 10.1 on the server so that I cannot install the latest version (i.e.pytorch-geometric 1.5.0+ ).

Great thanks!

**Comment by ezyang** [comment]
We're not going to do any more point releases of 1.4.0. Your best bet is building PyTorch from source for CUDA 10.1

**Comment by Wenhao-Jin** [comment]
OK, thanks fro your advice, I will try it!

Wenhao

**Comment by sunilkpai** [comment]
Hi @anjali411 I am getting that the backward op `indexing_backward` is not implemented for CUDA. I'm wondering if this falls under your checklist above? If not, I can create an issue!

**Comment by anjali411** [comment]
> Hi @anjali411 I am getting that the backward op `indexing_backward` is not implemented for CUDA for complex types. I'm wondering if this falls under your checklist above? 

@sunilkpai thanks for reaching out! yeah it should be added to the list above. 

> If not, I can create an issue!

that would be great! also make sure to link the issue in this umbrella issue

**Comment by jmlemercier** [comment]
Hi, I saw you ticked `torch.bmm` for CPU Complex Tensors, which is great, thanks ! Just wondering when the newer version implementing should be released, if you have an estimation.
Thanks a lot !

**Comment by anjali411** [comment]
> Hi, I saw you ticked `torch.bmm` for CPU Complex Tensors, which is great, thanks ! Just wondering when the newer version implementing this should be released, if you have an estimation.
> Thanks a lot !

@jmlemercier it exists on the master and will be released in 1.8 release in October.

**Comment by jmlemercier** [comment]
> > Hi, I saw you ticked `torch.bmm` for CPU Complex Tensors, which is great, thanks ! Just wondering when the newer version implementing this should be released, if you have an estimation.
> > Thanks a lot !
> 
> @jmlemercier it exists on the master and will be released in 1.8 release in October.

Great, thanks for the notice

**Comment by eliegenois** [comment]
Hi, having `torch.einsum` available on CUDA for Complex Tensors would be very helpful for my applications. It gets down to the error
> RuntimeError: _th_bmm_out not supported on CUDAType for ComplexDouble
I don't know if this feature will be added together with the point `torch.bmm [CUDA, waiting on TH to ATen port]`, but it would be great to have it. Do you have an estimation as to when this support will be added?
Thank you!

**Comment by eliegenois** [comment]
> @eliegenois Can you show a complete example that does not work?
> Did you try pytorch nightly? In 1.8.0.dev20201022+cpu the following works:
> 
> ```
> torch.einsum(
>     'ijk,ikl->ijl',
>     torch.zeros([2, 2, 2], dtype=torch.complex128),
>     torch.zeros([2, 2, 2], dtype=torch.complex128))
> ```

Thank you for your quick reply. Yes, I have pytorch nightly `1.8.0.dev20201022` and your command works as well, but the following does not (on the CUDA now):

> torch.einsum(
>     'ijk,ikl->ijl',
>     torch.zeros([2, 2, 2], dtype=torch.complex128, device='cuda:0'),
>     torch.zeros([2, 2, 2], dtype=torch.complex128, device='cuda:0'))

as I get the following complete error message

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-4d02ae6990e9> in <module>
----> 1 torch.einsum(
      2     'ijk,ikl->ijl',
      3     torch.zeros([2, 2, 2], dtype=torch.complex128, device='cuda:0'),
      4     torch.zeros([2, 2, 2], dtype=torch.complex128, device='cuda:0'))

/opt/miniconda3/envs/dyson-env/lib/python3.8/site-packages/torch/functional.py in einsum(equation, *operands)
    342         return einsum(equation, *_operands)
    343 
--> 344     return _VF.einsum(equation, operands)  # type: ignore
    345 
    346 

RuntimeError: _th_bmm_out not supported on CUDAType for ComplexDouble
```
Thank you for your help in solving this, it is much appreciated.

**Comment by zzxihuanheixiu** [comment]
Please help me！！！

The error information is as follows:
RuntimeError: eig_ backward: Backward calculation does not support complex eigenvalues at the moment.

The details are as follows:
One step in my loss function is to calculate the eigenvalues of the matrix, which I used torch.eig Function, but later when I reverse the derivation of the loss function, the program got stuck torch.eig The position of the function, because sometimes the matrix has complex eigenvalues, torch.eig Function does not support reverse derivation! How can I solve it???

**Comment by v0dro** [comment]
Can we add complex number support to `at::nan_to_num` too?

**Comment by anjali411** [comment]
> nan_to_num

yeah I believe so. could you file an issue requesting this functionality?

**Comment by mruberry** [comment]
> > nan_to_num
> 
> yeah I believe so. could you file an issue requesting this functionality?

Extending `nan_to_num()` to handle complex values sounds good, but I'd like to then propose we declare any complex number with a real and/or complex NaN as NaN, and we follow through on this decision in other operations, like with the `equal_nan` kwarg on `isclose()` and `TestCase.assertEqual()` and the new `testing.assert_close()`. For the latter two we already have "relaxed" NaN handling, and this update would make that unnecessary.

**Comment by gchhablani** [comment]
There is one more method that should support complex numbers - `torch.einsum`.

I am trying to port a code that uses complex numbers in einsum from JAX. Note that JAX has support for einsum on complex numbers. An example in JAX goes like this:

```python
>>> import jax.numpy as jnp
>>> from scipy.linalg import dft
>>> a = jnp.array([1,4,5,7,8])
>>> dft_matrix = dft(a.shape[0])
>>> dft_matrix
array([[ 1.        +0.j        ,  1.        +0.j        ,
         1.        +0.j        ,  1.        +0.j        ,
         1.        +0.j        ],
       [ 1.        +0.j        ,  0.30901699-0.95105652j,
        -0.80901699-0.58778525j, -0.80901699+0.58778525j,
         0.30901699+0.95105652j],
       [ 1.        +0.j        , -0.80901699-0.58778525j,
         0.30901699+0.95105652j,  0.30901699-0.95105652j,
        -0.80901699+0.58778525j],
       [ 1.        +0.j        , -0.80901699+0.58778525j,
         0.30901699-0.95105652j,  0.30901699+0.95105652j,
        -0.80901699-0.58778525j],
       [ 1.        +0.j        ,  0.30901699+0.95105652j,
        -0.80901699+0.58778525j, -0.80901699-0.58778525j,
         0.30901699-0.95105652j]])
>>> dft_matrix.shape
(5, 5)
>>> jnp.einsum('ij,j->i', dft_matrix, a)
DeviceArray([25.+0.j        , -5.+4.9797964j , -5.+0.44902802j,
             -5.-0.44902802j, -5.-4.9797964j ], dtype=complex64)
```

**Comment by mruberry** [comment]
> There is one more method that should support complex numbers - `torch.einsum`.
> 
> I am trying to port a code that uses complex numbers in einsum from JAX. Note that JAX has support for einsum on complex numbers. An example in JAX goes like this:
> 
> ```python
> >>> import jax.numpy as jnp
> >>> from scipy.linalg import dft
> >>> a = jnp.array([1,4,5,7,8])
> >>> dft_matrix = dft(a.shape[0])
> >>> dft_matrix
> array([[ 1.        +0.j        ,  1.        +0.j        ,
>          1.        +0.j        ,  1.        +0.j        ,
>          1.        +0.j        ],
>        [ 1.        +0.j        ,  0.30901699-0.95105652j,
>         -0.80901699-0.58778525j, -0.80901699+0.58778525j,
>          0.30901699+0.95105652j],
>        [ 1.        +0.j        , -0.80901699-0.58778525j,
>          0.30901699+0.95105652j,  0.30901699-0.95105652j,
>         -0.80901699+0.58778525j],
>        [ 1.        +0.j        , -0.80901699+0.58778525j,
>          0.30901699-0.95105652j,  0.30901699+0.95105652j,
>         -0.80901699-0.58778525j],
>        [ 1.        +0.j        ,  0.30901699+0.95105652j,
>         -0.80901699+0.58778525j, -0.80901699-0.58778525j,
>          0.30901699-0.95105652j]])
> >>> dft_matrix.shape
> (5, 5)
> >>> jnp.einsum('ij,j->i', dft_matrix, a)
> DeviceArray([25.+0.j        , -5.+4.9797964j , -5.+0.44902802j,
>              -5.-0.44902802j, -5.-4.9797964j ], dtype=complex64)
> ```
> 
> It would be great to have `torch.einsum` support complex numbers.

cc @heitorschueroff, who's been working on einsum

**Comment by BouazzaSE** [comment]
Referencing #65480 as it shows a case where `@torch.jit.script` fails to construct the autograd graph when working on complex tensors. A workaround for now is tracing.

**Comment by 0jg** [comment]
@gchhablani @mruberry @heitorschueroff (and anyone else!)

I have been trying to use `torch.einsum` on the complex domain, but it always returns the error on line 325 of `/torch/functional.py` (`325     return _VF.einsum(equation, operands)  # type: ignore[attr-defined]`)

`RuntimeError: expected scalar type ComplexDouble but found ComplexFloat`

I have checked all my data types, and they all seem to be `torch.complex128` (I assume "ComplexDouble"?) Is `einsum` generally incompatible with complex numbers, even on the nightly?

**Comment by mruberry** [comment]
@0jg would you post a small script that we can run and reproduce the error? It seems to be working with complex numbers OK for me.

**Comment by 0jg** [comment]
> @0jg would you post a small script that we can run and reproduce the error? It seems to be working with complex numbers OK for me.

This is now fixed! 

The matrices I had in my `einsum` (for a batched matrix multiplication) were a complex double and a complex float. The latter data type was hidden deep in the depths of my code.

**Comment by Lezcano** [comment]
Similar to the OpInfo issue, it might be good to declare victory with this one as well and simply close it? @mruberry @anjali411

**Comment by Ajk4** [comment]
Can we add to the list log1p and logcumsumexp? I see those missing and in my use case I actually need those for complex numbers.

Logcumsumexp for complex looks similar like double implementation, but in scan-loop we need to select min/max based on real part of complex. More context here https://scicomp.stackexchange.com/a/34275

**Comment by Lezcano** [comment]
I do not think there is a simple way to add complex support `log1p` as the standard library in C++ [does not support](https://en.cppreference.com/w/cpp/numeric/math/log1p) it either, and we rely on it for our support.

For this reason, I don't think we can implement a reasonably stable complex version of `logcumsumexp`, as it uses `log1p` internally
https://github.com/pytorch/pytorch/blob/6642e88ad23bfcf1349f2bce06f3490e49f195bf/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L116-L134

**Comment by Ajk4** [comment]
About log1p it might not be hopeless. I noticed some abandoned PR from year ago implementing it. 

LINK: https://github.com/pytorch/pytorch/pull/38465

**Comment by Lezcano** [comment]
@Ajk4 Alas, I don't have the time to work on this. I would suggest that you open a new issue describing the problem and linking the previous `log1p` PR and the discussion there.
In particular, if you wanted to have a stab at it, we'd be more than open to accept a contribution on this area.

**Comment by anjali411** [comment]
Closing this issue since we have added most of the listed support for complex numbers. Other requests can be handled in individual issues.
